### PR TITLE
[Core] Reduce node and dof size stage 2

### PR DESCRIPTION
**Pull request by pooyan-dadvand**
This PR changes the `Dof::Pointer` to raw pointer and also changes the Dof container in the node to `std::vector<Dof*>` 

It changes the behavior as the Dofs are not sorted anymore in each node. It also changes the API as now the Dof vector iterator points to the unique_ptr of Dof

@KratosMultiphysics/structural-mechanics I found several` dofs.Sort()` in your application which I had to remove. Please check that part of the code. (Tests are running in my machine)

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_strategies/builder_and_solvers/residualbased_block_builder_and_solver_periodic.h [hunk 1]
```diff
@@ -299,7 +299,7 @@ class ResidualBasedBlockBuilderAndSolverPeriodic
                         ModelPart::NodeType& rDest)
     {
         for (auto itDof = rOrigin.GetDofs().begin(); itDof != rOrigin.GetDofs().end(); itDof++)
-            rDest.pGetDof( itDof->GetVariable() )->SetEquationId( itDof->EquationId() );
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_strategies/builder_and_solvers/residualbased_block_builder_and_solver_periodic.h:302]: as a note for @jcotela i would say that this file should be eventually removed as it is superseded by the MPCs

File: kratos/includes/node.h [hunk 2]
```diff
@@ -407,7 +397,7 @@ class Node : public Point, public Flags
         // Deep copying the dofs
         for(typename DofsContainerType::const_iterator it_dof = rOther.mDofs.begin() ; it_dof != rOther.mDofs.end() ; it_dof++)
         {
-            pAddDof(*it_dof);
+            pAddDof(**it_dof);
```
Comment (RiccardoRossi) [inline kratos/includes/node.h:400]: i am not clear about this. what are u doing here? 

not telling it is wrong, just that i am not sure of what you are trying to accomplish (or better said, what the assignement operator for a node should do)

**Review by RiccardoRossi — COMMENTED** [review]
looks good from my side. 

i have some open questions, but they do not really depend on the current PR but rather are philosophical

waiting to approve to see if someone wants to make comments

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp [hunk 3]
```diff
@@ -672,19 +672,20 @@ class EigensolverStrategy
...
-            {
-                NodeDofs.Sort();
-            }
+            // TO BE VERIFIED!! In the current implmentation of Dofs there are nor reordered and only pushec back. 
+            // // the jth column index of EIGENVECTOR_MATRIX corresponds to the jth nodal dof. therefore,
+            // // the dof ordering must not change.
+            // if (NodeDofs.IsSorted() == false)
+            // {
+            //     NodeDofs.Sort();
+            // }
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:687]: @msandre I think you initially wrote this, can you please have a look?

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp [hunk 4]
```diff
@@ -263,19 +263,20 @@ class HarmonicAnalysisStrategy
...
 
-                    if( node_dofs.IsSorted() == false )
-                    {
-                        node_dofs.Sort();
-                    }
+                    // TO BE VERIFIED!! In the current implmentation of Dofs there are nor reordered and only pushec back. 
+                    // if( node_dofs.IsSorted() == false )
+                    // {
+                    //     node_dofs.Sort();
+                    // }
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp:274]: @qaumann please have a look, but I guess this comes from the EigensolverStrategy

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp [hunk 5]
```diff
@@ -428,15 +429,16 @@ class HarmonicAnalysisStrategy
...
 
-                if (node_dofs.IsSorted() == false)
-                {
-                    node_dofs.Sort();
-                }
+            // TO BE VERIFIED!! In the current implmentation of Dofs there are nor reordered and only pushec back. 
+                // if (node_dofs.IsSorted() == false)
+                // {
+                //     node_dofs.Sort();
+                // }
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp:436]: same

**Comment by loumalouomega** [comment]
I fixed the contact and the meshing application

File: kratos/includes/node.h (hunk 2 above)
Comment (pooyan-dadvand) [reply kratos/includes/node.h:400]: Before this PR the *it_dof was a reference to a `Dof`. Now it is a reference to the `unique_ptr<Dof>`. So I need another star to get the reference to the `Dof`.

Comment (RiccardoRossi) [reply kratos/includes/node.h:400]: then you might want to use

(*it_dof).get()

in any case thx for the explanation and feel free to ignore the suggestion

**Comment by pooyan-dadvand** [comment]
Thanks, @loumalouomega!

**Comment by RiccardoRossi** [comment]
on my side i did a little fix to allow compilation of the mpi stuff. it would be good that someone takes a look at that too

**Comment by pooyan-dadvand** [comment]
I am adding a sort step to AddDofs so they are stored in order of their keys as before. 

First I wanted to avoid this overhead but I realize that in many places we are accessing the dofs by their array index and having them in different orders in different nodes can cause serious problem.

I hope this would also fix the eigenvalue tests

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp (hunk 3 above)
Comment (msandre) [reply applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp:687]: I believe it is ok as long as the dof ordering is fixed. You could use std::sort if you don't want to change the behavior.

**Comment by jcotela** [comment]
@RiccardoRossi the MPI changes look fine to me.

**Comment by philbucher** [comment]
the trilinosapp does not compile
I am trying to fix it

**Comment by philbucher** [comment]
the trilinosapp is fixed in #5567

**Comment by pooyan-dadvand** [comment]
I just made a test, allocating 1M nodes with 6 Dofs per each node and we reduced from 994MB before any reduction to 623MB in this stage! :sunglasses:

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/pfem 

I have realized that in several places you are copying the Dofs vector from one node to the other. Some examples are:

https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemApplication/custom_processes/insert_fluid_nodes_mesher_process.hpp#L357

https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemApplication/custom_processes/insert_fluid_nodes_mesher_process.hpp#L359

https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L377

https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L379

https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L410

This way of copying dofs may cause serious problems in accessing the data structure afterward and will not be possible anymore with the changes in this PR.

Would you please check these parts and try to copy the Dof one by one instead of copying the pointer?

**Update:**
I have just reimplemented them doing a deep copy of Dofs. Please verify if this behavior was the intention behind the code.

**Comment by ipouplana** [comment]
> @KratosMultiphysics/pfem
> 
> I have realized that in several places you are copying the Dofs vector from one node to the other. Some examples are:
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemApplication/custom_processes/insert_fluid_nodes_mesher_process.hpp#L357
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemApplication/custom_processes/insert_fluid_nodes_mesher_process.hpp#L359
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L377
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L379
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L410
> 
> This way of copying dofs may cause serious problems in accessing the data structure afterward and will not be possible anymore with the changes in this PR.
> 
> Would you please check these parts and try to copy the Dof one by one instead of copying the pointer?

Hi @pooyan-dadvand , did you already fix it with eb46d6d ?

**Comment by pooyan-dadvand** [comment]
and in `3072743`

**Comment by pooyan-dadvand** [comment]
But I'm not sure if the fix is what was intended

**Comment by ipouplana** [comment]
> @KratosMultiphysics/pfem
> 
> I have realized that in several places you are copying the Dofs vector from one node to the other. Some examples are:
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemApplication/custom_processes/insert_fluid_nodes_mesher_process.hpp#L357
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemApplication/custom_processes/insert_fluid_nodes_mesher_process.hpp#L359
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L377
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L379
> 
> https://github.com/KratosMultiphysics/Kratos/blob/23de0dc83805f032f8a2c53ce6992187c62d9a19/applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp#L410
> 
> This way of copying dofs may cause serious problems in accessing the data structure afterward and will not be possible anymore with the changes in this PR.
> 
> Would you please check these parts and try to copy the Dof one by one instead of copying the pointer?
> 
> **Update:**
> I have just reimplemented them doing a deep copy of Dofs. Please verify if this behavior was the intention behind the code.

Thanks @pooyan-dadvand 

@AFranci could you look into this? I cannot checkout to another branch right now...

**Comment by AFranci** [comment]
I have just checked and everything seems to be ok! Thank you @pooyan-dadvand for the changes in the PfemFluidApplication

**Comment by pooyan-dadvand** [comment]
It would be great if someone approves it!

**Review by philbucher — CHANGES_REQUESTED** [review]
@pooyan-dadvand this breaks the TrilinosApp
=> #5567

**Review by philbucher — APPROVED** [review]

**Pull request merged**